### ISSUE TITLE
Can't make PDF with Dynamic Data

### ISSUE THREAD

**Issue by DevQueen** [bpampuch/pdfmake#454]
I wish someone would post some examples of using dynamic data in Angular to generate a pdf.  I found most of this code on the site but can't get it to work with data from Angular Service.  Currently, I receive the following error when I try to run the code:

TypeError: Object doesn't support property or method 'forEach'
   at buildTableBody (http://localhost/SSQV4/SSQV5/Scripts/app/Contractor/ContractorCtrl.js:55:21)
   at table (http://localhost/SSQV4/SSQV5/Scripts/app/Contractor/ContractorCtrl.js:69:21)
   at $scope.export (http://localhost/SSQV4/SSQV5/Scripts/app/Contractor/ContractorCtrl.js:77:17)
   at fn (Function code:2:277)
   at Anonymous function (http://localhost/SSQV4/SSQV5/Scripts/angular-touch.js:478:9)
   at Scope.prototype.$eval (http://localhost/SSQV4/SSQV5/Scripts/angular.js:15989:9)
   at Scope.prototype.$apply (http://localhost/SSQV4/SSQV5/Scripts/angular.js:16089:13)
   at Anonymous function (http://localhost/SSQV4/SSQV5/Scripts/angular-touch.js:477:7)
   at jQuery.event.dispatch (http://localhost/SSQV4/SSQV5/Scripts/jquery-1.8.2.js:3076:6)
   at eventHandle (http://localhost/SSQV4/SSQV5/Scripts/jquery-1.8.2.js:2694:5)

Here is the code:

```
       var externalDataRetrievedFromServer = $scope.scoredetails;

            function buildTableBody(data, columns) {
                var body = [];

                body.push(columns);

                data.forEach(function (row) {
                    var dataRow = [];

                    columns.forEach(function (column) {
                        dataRow.push(row[column].toString());
                    })

                    body.push(dataRow);
                });

                return body;
            }

            function table(data, columns) {
                return {
                    table: {
                        headerRows: 1,
                        body: buildTableBody(data, columns)
                    }
                };
            }

            var dd = {
                content: [
                    { text: 'Dynamic parts', style: 'header' },
                    table(externalDataRetrievedFromServer, [ 'AddedPoints',
    'BusinessUnitID','dteDateCalculated','intMaxPossibleScore','intNumericalScore',
    'OverRidden','ScoringSetID','ServiceRegionID','ServiceRegionName',
   'SrvRegOrder','vchCurrentOperatorRanking','vchCurrentRankingColor',
   'vchScoreOverridenByUserName', 'vchScoringSetName'])
                ]
            };

            pdfMake.createPdf(dd).download();
```

This code works if I type in "static" data in the variable "externalDataRetrievedFromServer".

Here is what my data looks like when retrieved from the server:

![pdfmake error](https://cloud.githubusercontent.com/assets/13017434/11594321/f15d5b60-9a6e-11e5-8cbf-41903fa8ab40.png)

I really need some help as not being able to generate a PDF from my Angular controller is holding up my project.  ANY assistance is greatly appreciated!

**Comment by ljfolino** [comment]
Not sure if this will work fro you but I had a similar issue.  Some of the observations I had when crating dynamic data table was that I had to have exactly the same number of elements peer object.  This meant converting nulls to empty strings.  I created a helper function to do this that basically checked nulls.  

The second issue I found was that I couldn't include the column headers with the array in the doc definition.  What I ended up doing was adding the headers to the array instead of including both in the doc definition.

Like this
outside doc def
var items = getFromRemote();
items.unshift(header);

in doc definition
                  table:
                            {
                                headerRows: 1,
                                widths: ['7%', '7%', '*', '7%', 'auto', '5%', '7%', '7%', '7%', 'auto', '*'],
                                body: items

                            },
